src/SarcophagusApi.ts

```diff
@@ -283,39 +286,81 @@ export class SarcophagusApi {
 
       let payload: { type: string; data: Buffer };
       let fileName: string;
+
+      let keySharesEncryptedInner: Uint8Array[]
+      let encryptedPayload: Buffer;
+
+      let doInnerEncryption = true;
+
+      // The caller may provide pre-encrypted keyshares and a pre-encrypted payload.
+      // In this case, we skip the inner encryption step and use the provided encrypted keyshares and payload.
+      if (preEncryptedPayload && !recipientInnerEncryptedkeyShares) {
+        throw new Error('`preEncryptedPayload` provided but `recipientInnerEncryptedkeyShares` not provided');
+      }
+      if (recipientInnerEncryptedkeyShares && !preEncryptedPayload) {
+        throw new Error('`recipientInnerEncryptedkeyShares` provided but `preEncryptedPayload` not provided');
+      }
+
+      if (!recipientInnerEncryptedkeyShares && !preEncryptedPayload && !file && !payloadData) {
+        throw new Error("Can't upload file. No file, payload, or inner-encrypted data provided.");
+      }
+
+      if (recipientInnerEncryptedkeyShares && preEncryptedPayload) {
+        keySharesEncryptedInner = recipientInnerEncryptedkeyShares.map(arr => Buffer.from(arr));
+        encryptedPayload = Buffer.from(preEncryptedPayload);
+        doInnerEncryption = false;
+      }
+
+      if (!doInnerEncryption && !preEncryptedPayloadMetadata) {
+        throw new Error('`preEncryptedPayloadMetadata` must be provided when `recipientInnerEncryptedkeyShares` and `preEncryptedPayload` are provided');
+      }
+      
+      if ((file || payloadData) && !doInnerEncryption) {
+        throw new Error('Provide only one of `file`, `payloadData` or `recipientInnerEncryptedkeyShares` and `preEncryptedPayload`.');
+      }
```

Comment:
I think it might be better to split out the 2 phases of encryption to separate helper methods, and then:
1. Create a separate public API method for single-encryption that uses the helper methods
2. Refactor this current method to use the encryption helper methods

Having 2 separate public methods will scope down the responsibility of each. The args can also be typed as required for this second method so you shouldn't have to check whether they exist or not.